src/d.sea-of-nodes.md

```diff
@@ -0,0 +1,327 @@
+---
+title: Sea of Nodes
+date: 2015-10-08
+---
+
+## Outline
+
+1. Compilers are translating
+2. Text, AST - not ok for optimizations, need for data-flow graph
+3. Usually control graph is totally separated from data-flow graph, CFG
+4. Sea-of-nodes - same graph, two kinds of edges
+5. Graph reductions
+6. Reachability + const prop example
+7. Scheduling stuff back in blocks
+8. Generating machine code
+
+## Brief intro
+
+This post is going to be about the sea-of-nodes compiler concept that I have
+recently learned.
+
+While it is completely not necessary, it may be useful to take a peek at the
+some of my previous posts on JIT-compilers before reading this:
+
+* [How to start JIT-ting][0]
+* [Allocating numbers][1]
+* [SMIs and Doubles][2]
+* [Deoptimize me not, v8][3]
+
+## Compilers = translators
+
+Compilers is something that every Software Engineer uses several times a day.
+Surprisingly even people who consider themselves to be far from writing the
+code, still use the compilers quite heavily throughout their day. This is
+because most of the web depends on client-side code execution, and many of such
+client-side programs are passed to the browser in a form of the source code.
+
+Here we come to important thing: while source code is (usually) human-readable,
+it looks like a complete garbage to your laptop/computer/phone/...'s CPU. On
+other hand, machine code, that computers **can** read, is (almost always) not
+human-readable. Something should be done about it, and the solution to this
+problem is provided by the process called **translation**.
+
+Trivial compilers do a single pass of _translation_: from the source code to the
+machine code. More usual compilers do at least two passes: from the source code
+to [Abstract Syntax Tree][4] (AST), and from [AST][4] to machine code. [AST][4]
+in this case acts like an _Intermediate Representation_ (IR), and as the name
+suggests, [AST][4] is just another form of the same source code.
+
+There is no limit on the layer count. Each new layer brings the source program
+closer to how it will look like in machine code.
+
+## Optimization layers
+
+However, not all layers are used just for translation. Many compilers are also
+try to optimize the human-written code. (There is always a balance between code
+elegance and code performance).
+
+Take the following JavaScript code, for example:
+
+```javascript
+for (var i = 0; i < arr.length; i++)
+  arr[i] += 1;
+```
+
+If the compiler would translate it to the machine code straight out of [AST][4],
```

Comment:
```diff
- If the compiler would translate it to the machine code straight out of [AST][4], it may look this (in very abstract and detached from reality instruction set):
+ If the compiler naively translated the machine code directly from the [AST][4], it may resemble the following output (in a very abstract and detached from reality instruction set):
```